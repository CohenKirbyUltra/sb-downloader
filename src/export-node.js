import environment from './environment.js';

environment.canAccessScratchAPI = true;

// The version here should be incremented if our traffic pattern ever changes significantly
environment.headers['user-agent'] = 'SBDL/1.0 (+https://www.npmjs.com/package/@turbowarp/sbdl)';

export * from './downloader.js';
