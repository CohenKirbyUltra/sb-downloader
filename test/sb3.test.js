import fs from 'fs';
import * as SBDL from '../src/export-node.js';
import JSZip from 'jszip';
import {getFixturePath, arrayBufferSerializer} from './test-utilities.js';
import {expect, test} from 'vitest';

expect.addSnapshotSerializer(arrayBufferSerializer);

test('sb3 project from JSON', async () => {
  const fixture = getFixturePath('sb3-167118244.json');
  const project1 = await SBDL.downloadProjectFromJSON(fs.readFileSync(fixture, 'utf-8'));
  const project2 = await SBDL.downloadProjectFromJSON(JSON.parse(fs.readFileSync(fixture, 'utf-8')));
  const project3 = await SBDL.downloadProjectFromBuffer(fs.readFileSync(fixture));
  expect(project1.type).toBe('sb3');
  expect(project1.title).toBe('');
  expect(project1.arrayBuffer).instanceOf(ArrayBuffer);
  // All methods of loading the project should result in identical data
  expect(project1).toStrictEqual(project2);
  expect(project2).toStrictEqual(project3);
  // The data should not change unexpectedly
  expect(project1).toMatchSnapshot();
}, 30000);

test('sb3 project from sb3', async () => {
  const fixture = getFixturePath('167118244.sb3');
  const originalData = fs.readFileSync(fixture);
  const project = await SBDL.downloadProjectFromBuffer(originalData);
  expect(project.type).toBe('sb3');
  expect(project.title).toBe('');
  expect(new Uint8Array(project.arrayBuffer)).toStrictEqual(new Uint8Array(originalData.buffer));
});

test('sb3 project from sb3 with project.json in a subdirectory', async () => {
  const fixture = getFixturePath('json-in-subdirectory.sb3');
  const originalData = fs.readFileSync(fixture);
  const project = await SBDL.downloadProjectFromBuffer(originalData);
  expect(project.type).toBe('sb3');
  const zip = await JSZip.loadAsync(project.arrayBuffer);
  expect(zip.file('this is a subdirectory/project.json')).toBeNull();
  expect(zip.file('this is a subdirectory/9838d02002d05f88dc54d96494fbc202.png')).toBeNull();
  expect(zip.file('project.json')).not.toBeNull();
  expect(zip.file('9838d02002d05f88dc54d96494fbc202.png')).not.toBeNull();
  expect(project.arrayBuffer).toMatchSnapshot();
});

test('downloads missing assets', async () => {
  const rootFixtureData = fs.readFileSync(getFixturePath('missing-assets-root.sb3'));
  let loadedAssets = -1;
  let totalAssets = -1;
  const project1 = await SBDL.downloadProjectFromBuffer(rootFixtureData, {
    onProgress: (type, loaded, total) => {
      if (type === 'assets') {
        loadedAssets = loaded;
        totalAssets = total;
      }
    }
  });
  expect(loadedAssets).toBe(3);
  expect(totalAssets).toBe(3);
  expect(project1.arrayBuffer).toMatchSnapshot();

  const subdirectoryFixtureData = fs.readFileSync(getFixturePath('missing-assets-subdir.sb3'));
  const project2 = await SBDL.downloadProjectFromBuffer(subdirectoryFixtureData);
  expect(project2.arrayBuffer).toMatchSnapshot();

  // Both of the fixtures are the same project, just one of them has some files in subdirectories
  // The actual order of the files in the zip is not necessarily the same (so they may not be 
  // the same SHA-256) but the contents in those files should be.

  const zip1 = await JSZip.loadAsync(project1.arrayBuffer);
  const zip2 = await JSZip.loadAsync(project2.arrayBuffer);
  expect(Object.keys(zip1).sort()).toStrictEqual(Object.keys(zip2).sort());
  for (const path of Object.keys(zip1.files)) {
    expect(await zip1.file(path).async('uint8array')).toStrictEqual(await zip2.file(path).async('uint8array'));
  }
});
