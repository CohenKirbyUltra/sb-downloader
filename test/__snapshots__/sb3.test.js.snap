// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`downloads missing assets 1`] = `ArrayBuffer [SHA-256 bbded48ad46f4c1fa9929ffadd17ecdfa94b2d322c8a7e0409c629c44d103f41]`;

exports[`downloads missing assets 2`] = `ArrayBuffer [SHA-256 4f25165fae50acf1ae0c05e19e0ecc363e3c93fb263ad84c67b1412ab53e5a4c]`;

exports[`extra empty subdirectory is removed 1`] = `ArrayBuffer [SHA-256 4532296b54eb844a047bf07efa5ff4a21fcff07df64a91a344ca1a7a4a18a8fe]`;

exports[`sb3 project from JSON 1`] = `
{
  "arrayBuffer": ArrayBuffer [SHA-256 73f90aff8695ab18da8736f7a0b5054bed29cb2fc08fe59ab34da7d82c6a74d0],
  "title": "",
  "type": "sb3",
}
`;

exports[`sb3 project from sb3 with project.json in a subdirectory 1`] = `ArrayBuffer [SHA-256 232be6275f7745ab2730e9fa3d83d6417b9f8ed4c05e5d5e38453301ff2d4840]`;
